src/tree.rs

```diff
@@ -247,51 +306,114 @@ where
-            }
 
-            NodeType::Branch(ref branch) => {
-                let child = branch.find_child(point, self.nodes[node].aabb.center());
-                self.rfind(child, point)
-            }
+                NodeType::Branch(ref branch) => {
+                    node = branch.find_child(point, self.nodes[node].aabb.center());
+                    continue;
+                }
+            };
+        }
+    }
+
+    /// Returns the node's [`id`](NodeId) containing the element if element exists and not garbaged.
+    pub fn get_node_id(&self, element: ElementId) -> Option<NodeId> {
+        if self.map.is_garbaged(element) {
+            None
+        } else {
+            Some(self.map[element.into()])
         }
     }
 
     /// Returns the node's [`id`](NodeId) containing the element if element exists and not garbaged.
-    pub fn get_node(&self, element: ElementId) -> Option<NodeId> {
+    pub fn get_node(&self, element: ElementId) -> Option<Node<U>> {
         if self.map.is_garbaged(element) {
             None
         } else {
-            Some(self.map[element])
+            Some(self.nodes[self.map[element.into()]])
+        }
+    }
+
+    /// Returns the element if element exists and not garbaged.
+    pub fn get_element(&self, element: ElementId) -> Option<&T> {
+        if self.elements.is_garbaged(element) {
+            None
+        } else {
+            Some(&self.elements[element])
         }
     }
 
     /// Consumes a tree, converting it into a [`vector`](Vec).
     pub fn to_vec(self) -> Vec<T> {
-        let garbage = self.elements.garbage;
         self.elements
             .vec
             .into_iter()
-            .enumerate()
-            .filter_map(|(i, element)| {
-                if garbage.contains(&i) {
-                    None
-                } else {
-                    Some(element)
-                }
-            })
+            .filter(|e| !e.garbage)
+            .map(|e| e.item)
             .collect()
     }
+
+    /// Returns the number of actual elements in the tree
+    pub fn len(&self) -> usize {
+        self.elements.len()
+    }
+
+    /// Is the tree empty
+    pub fn is_empty(&self) -> bool {
+        self.elements.is_empty()
+    }
+
+    /// Returns an iterator over the elements in the tree.
+    pub fn iter(&self) -> PoolIterator<'_, T> {
+        self.elements.iter()
+    }
+
+    /// Returns an iterator over the nodes in the tree.
+    pub fn iter_nodes(&self) -> PoolIterator<Node<U>> {
+        self.nodes.iter()
+    }
+
+    /// Returns an iterator over the elements in the tree.
+    pub fn iter_elements(&self) -> PoolElementIterator<'_, T> {
+        self.elements.iter_elements()
+    }
+}
+
+impl<U: Unsigned, T: Position<U = U>> std::iter::IntoIterator for Octree<U, T> {
+    type Item = T;
+    type IntoIter = PoolIntoIterator<T>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.elements.into_iter()
+    }
+}
+
+impl<U, T> Clone for Octree<U, T>
```

Comment:
Gonna replace that with derive macro too?
```suggestion
#[derive(Clone)]
```

Reply:
Yeah, this is what I feared, I'll need to undo macros and go back to explicit

```shell
   Compiling battle-cities v0.1.0 (C:\Users\johna\prog\battle-cities\crates\client)
error[E0277]: the trait bound `dynamic_game_world::update_world_visibility::WorldVisibilityNode: Clone` is not satisfied
  --> crates\client\src\controller\dynamic_game_world\update_world_visibility.rs:20:5
   |
18 | #[derive(Component, Clone)]
   |                     ----- in this derive macro expansion
19 | pub struct WorldVisibility {
20 |     pub octree: Octree<u64, WorldVisibilityNode>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `dynamic_game_world::update_world_visibility::WorldVisibilityNode`, which is required by `Octree<u64, dynamic_game_world::update_world_visibility::WorldVisibilityNode>: Clone`
   |
   = note: required for `Octree<u64, dynamic_game_world::update_world_visibility::WorldVisibilityNode>` to implement `Clone`
   = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider annotating `dynamic_game_world::update_world_visibility::WorldVisibilityNode` with `#[derive(Clone)]`
   |
7  + #[derive(Clone)]
8  | pub struct WorldVisibilityNode {
   |

For more information about this error, try `rustc --explain E0277`.
error: could not compile `battle-cities` (bin "battle-cities") due to 1 previous error
PS C:\Users\johna\prog\battle-cities>
```

Reply:
Last commit fixes the Clone requirement, not mandatory anymore